Integrated working PyTorch-CRF in MM

vrdn-23 opened the pull request:
Folks,

This is the first draft PR that I have ready for moving from the old rusty sklearn-crfsuite to the new (hopefully better) pytorch-CRF. A couple of things to note here:

- I haven't really removed the original CRF code from the Mindmeld repo just yet. I still want to make sure we don't have major accuracy drops when it comes to the ANLP repo. My individual experiments on the ANLP repo suggests that we should be fine, but there could always be pitfalls that I did not account for while integrating and I want to be sure before we move on to something new. Hence for now, that way this is structured is that we can switch between the cry models based on "crf" and "torch-crf". Ideally I would either want to remove the old CRF completely by the time this PR is merged or mark it as deprecated for a couple of RCs and refactor it out once we move to the next major release.
- I still need to add some unit tests to this, although I'm not sure if there are particular unit tests that would work for a PyTorch model
- The design for this isn't really perfect. So I would appreciate any suggestions or modifications for a much cleaner codebase. I'm trying to make sure that we stick to the crfsuite implementation that we have as closely as possible so that when we do eventually switch models, it doesn't involve a lot of refactoring and that most things are taken care of by the new PyTorch class. If this might be the wrong approach to look at things, please let me know!
- The functions for compute_marginals which is required for the predict_proba class in CRF has been [borrowed from a PR](https://github.com/kmkurn/pytorch-crf/pull/37) raised against the original pytorch-crf repo. For reasons unknown, it never got merged into the master branch, but I've gone through the implementation a bit and it looks solid and the original author also seemed to approve the implementation. 

Please feel free to comment or let me know if you have any questions regarding the PyTorch code, or if I've missed anything essential with regards to integrating this new monstrosity!
Thanks for all the help in advance! Looking forward to meeting everyone at a newer scikit-learn version soon! :)

snow0x2d0 commented:
![EEEEEEeeeeeEEEEE](https://media.giphy.com/media/5GoVLqeAOo6PK/giphy.gif)

vembar commented on mindmeld/models/taggers/pytorch_crf.py:
```diff
@@ -0,0 +1,419 @@
+
+
+class Encoder:
+    def __init__(self, feature_extractor="dict", num_feats=None):
```
Looks like we set the default value for feature_extractor as "dict" here, but a "hash" in the default config?

vrdn-23 replied:
Fixed and changed both to hash

vembar commented on mindmeld/models/taggers/pytorch_crf.py:
```diff
@@ -0,0 +1,419 @@
+        return torch.exp(prob).transpose(0, 1)
+
+    @staticmethod
+    def _log_sum_exp(tensor_input, dim):
```
Is this necessary? Can't we use torch.logsumexp() like in the previous method?
Documentation seems to suggest that it is numerically stable - https://pytorch.org/docs/stable/generated/torch.logsumexp.html#torch.logsumexp

vrdn-23 replied:
You're right! I left it in there cause the original author in the PR used it as such. But I ran some preliminary tests and it doesn't seem like it's any different behavior than what we expect from the in-built torch function!

vrdn-23 commented:
Folks,
Just a small update since this hasn't seem any activity from my side. I ran some of the experiments to determine runtime and memory consumption of the new CRF module on my local system last week and the results have been inconsistent based on what I've been running in the background. As such I moved the experiments to an online server this week, and they are being run as we speak. 

I have enough evidence to suggest that the accuracy is not too bad from whence I first ran the experiments and am convinced that this PR is ready to graduate from its "Draft" status. It would be great if everyone could let me know what you think looks good and what you think needs a complete overhaul. I will be giving my full active attention to the PR starting this Monday (05/16) and hope to have the time/memory/accuracy analysis done and updated on the PR by then. 

I hope to have all outstanding comments resolved by Monday EOD and looking forward to addressing new ones!
TIA!

vrdn-23 commented:
<img width="912" alt="Screen Shot 2022-05-19 at 15 34 28" src="https://user-images.githubusercontent.com/16606656/169415457-a233c47f-3425-451a-abb3-5fcb43231d4f.png">

I don't have a a better way of formatting this as a markdown table, so I'm going to leave this a picture but these are the main conditions I think are worth knowing about this data:
- These numbers are the averages across 15 runs. 
- The different configurations tried here are the original CRF (sklearn-crfsuite), the torch-crf using the "dict" (aka dictionary) configuration, the torch-crf using the "hash" (aka feature hasher) configuration with a patience of 3, the torch-crf "hash" with a patience of 5. All of them use a batch size of 10 and other defaults are kept unless mentioned otherwise.
- For the time column, the times shown are the total time taken to train and predict on the entire train and test dataset respectively.
- These are the only 3 intents that use CRF in ANLP and hence these were the only ones I ran these numbers on. 
- The memory and time measurements are based on the memory usage and time taken to run the "er.fit()" and "er.predict()" lines for each intent-specific entity recognizer. 

The current configuration that I would support for use in ANLP would be the "hash" configuration with patience 5. The rest of the experimental settings are included to showcase an overall picture of how the torch-crf performs.
Some key takeaways that can be made from these values are:
- We do see an increase in memory used by the torch-crf and it is highest when we use the "dict" feature extractor and this is expected as the dictionary feature extractor takes into account all possible features while the feature hasher, helps reduce the number of sparse features we have. There is a 50-100 MiB increase in memory while training in the other configurations and if this is a problem that needs to be addressed, I can look further into seeing if there is anything that I can do to reduce the memory footprint of the code. However as expected, there is no increase in memory in increasing the patience, so choosing the patience: 5 configuration does not hurt us in this case.
- The times for training are not as bad as I had initially anticipated. Apart from the "dict" configuration, at worst we are probably 3-4 minutes slower than training from the original CRF. Personally, I don't see how we can speed up this and reducing the patience obviously has a positive effect in improving the time it takes to train, but I think that also comes at the cost of the final accuracy. Hence in the case of the pytorch-crf, choosing faster train times might come at the cost of accuracy which is why I'm preferring having a longer patience time as opposed to a smaller one. 
- One good thing I noticed is that it seems for bigger datasets (esp call_person) it does seem like inference is happening much faster than when using the original CRF although the improvement might not be as obvious when comparing on smaller datasets.
- The accuracies are not really that bad in terms on the original model. Atmost we see a 1-2% drop in all configs, but the primary reason for this, is the fact that we are sacrificing 15% of our training data for validation and my initial experiments did show that we could match the accuracies we got with the original CRF setup if allowed to train with the full ANLP data. The best way to fix this would be to invest time into a dev set that we can rely on for ANLP.

I realize that this is a lot to take in. But if you have any questions or comments, please feel free to reach out to me 1-on-1 or drop a comment here! I'm hoping that with enough review comments, we can make this code a little bit more cleaner/faster/lighter or all of the above!

vrdn-23 commented:
![6gtal3](https://user-images.githubusercontent.com/16606656/169417956-44c45bcd-a2f9-4d7d-a7cf-547a4f1b736c.jpg)
cc @murali1996 @mattfrey-cisco @vembar @ritvikshrivastava @snow0x2d0 @tmehlinger @serapio

tmehlinger commented on mindmeld/models/taggers/pytorch_crf.py:
```diff
@@ -0,0 +1,412 @@
+        mask = torch.as_tensor(mask_list, dtype=torch.bool)
+        if self.labels:
+            return self.inputs[index], mask, self.labels[index]
+        else:
```
you don't need `else` if you return early above :)

vrdn-23 replied:
Gotcha! Fixed!

tmehlinger commented on mindmeld/models/taggers/pytorch_crf.py:
```diff
@@ -0,0 +1,412 @@
+        self.epochs = params.get('epochs', DEFAULT_PYTORCH_CRF_ER_CONFIG['epochs'])
+        self.train_dev_split = params.get('train_dev_split', DEFAULT_PYTORCH_CRF_ER_CONFIG['train_dev_split'])
+        self.optimizer_type = params.get('optimizer_type', DEFAULT_PYTORCH_CRF_ER_CONFIG['optimizer_type']).lower()
+        self.verbose = (logger.getEffectiveLevel() == logging.DEBUG)
```
You don't need to gate calls to `logger.debug` with a flag. Just emit debug logs and we can control output by setting the debug log level.

vrdn-23 replied:
Oof. Can't believe I didn't think of that sooner! Thanks a bunch!

vrdn-23 commented on mindmeld/models/taggers/crf.py:
```diff
@@ -182,6 +183,150 @@ def setup_model(self, config):
+    """A Conditional Random Fields model."""
+
+    @staticmethod
+    def _predict_proba(X):
```
Fixed. Thanks!

vrdn-23 commented on mindmeld/models/taggers/crf.py:
```diff
@@ -182,6 +183,150 @@ def setup_model(self, config):
+        pass
+
+    @staticmethod
+    def load(model_path):
```
Fixed again!

mattfrey-cisco commented:
@vrdn-23 - I don't think the memory usage matters too much anymore, @tmehlinger can weigh in on that.  When we were building for SPLAT, we had to use the env var documented here: https://www.mindmeld.com/docs/userguide/getting_started.html#mm-crf-features-in-memory

If you rerun the tests with that set to '0' the original CRF training will use significantly less memory.  I'm fine removing that option, but we need to verify our build pipelines can handle that memory usage, and we should remove that env var from the documentation.

vembar commented on mindmeld/models/taggers/pytorch_crf.py:
```diff
@@ -0,0 +1,407 @@
+DEFAULT_PYTORCH_CRF_ER_CONFIG = {
+    "feat_type": "hash",  # ["hash", "dict"]
+    "feat_num": 50000,
+    "stratify": True,
```
Nit: Can we rename to `stratify_train_val_split`?

vrdn-23 replied:
Fixed

vembar commented on mindmeld/models/taggers/pytorch_crf.py:
```diff
@@ -0,0 +1,407 @@
+    "train_batch_size": 8,
+    "patience": 3,
+    "epochs": 100,
+    "train_dev_split": 0.15,
```
similar nit: Can we rename to `train_val_split`? Don't want the dev set here to be confused with dev set files you can add as part of the hierarchy. Can be persuaded otherwise though :)

vrdn-23 replied:
Renamed to train_val _split. I do think that would be the better recognized variable name.

murali1996 replied:
Just a thought: It can be easier for MindMeld users/developers if params terminology is similar across different modules where pytorch is being used (see [params](https://www.mindmeld.com/docs/userguide/deep_neural_nets.html#common-configurable-params) used in deep NNs docs). Things like epochs/number_of_epochs, train_val_split/dev_split_ratio, train_batch_size/batch_size, etc.

vembar commented on mindmeld/models/taggers/pytorch_crf.py:
```diff
@@ -0,0 +1,407 @@
+        self.num_classes = None
+        self.classes = None
+        self.num_feats = num_feats
+        self.fit_done = False
```
To keep things consistent across the repo, can we change this to `self.ready`?

vrdn-23 replied:
Changed. Although I'm not sure I've done anything special to make sure the self.ready is inherited anywhere. This is just on the encoder side. Not the model side.

vrdn-23 replied:
Resolving conversation to shift it to under Lucien's comment.

vembar commented on mindmeld/models/taggers/pytorch_crf.py:
```diff
@@ -0,0 +1,407 @@
+                self.classes = self.label_encoder.classes_
+                self.num_classes = len(self.label_encoder.classes_)
+
+            self.fit_done = True
```
`self.ready`

vrdn-23 replied:
Changed

vembar commented on tests/models/test_tagging.py:
```diff
@@ -271,7 +271,7 @@ def test_get_boundary_counts_sequential(
 @pytest.mark.parametrize(
     "model_type,params",
-    [("memm", {"penalty": "l2", "C": 10000}), ("crf", {"c1": 0.01, "c2": 0.01})],
+    [("memm", {"penalty": "l2", "C": 10000}), ("crf", {"c1": 0.01, "c2": 0.01}), ("torch-crf", {"feat_type": "dict"})],
```
can we add a test for when feature type is hash?

vrdn-23 replied:
Done!

vrdn-23 commented on mindmeld/models/tagger_models.py:
```diff
@@ -27,7 +27,7 @@
 from .model import ModelConfig, Model, PytorchModel, AbstractModelFactory
 from .nn_utils import get_token_classifier_cls, TokenClassificationType
-from .taggers.crf import ConditionalRandomFields
+from .taggers.crf import ConditionalRandomFields, PyTorchCRF
```
I've changed it. I was struggling on how to make the distinction between Tagger, Model and Layer. Thanks a lot!

vrdn-23 commented on mindmeld/models/taggers/pytorch_crf.py:
```diff
@@ -0,0 +1,407 @@
+        return self.inputs[index], mask
+
+
+def custom_coo_cat(tensors):
```
Changed.

vrdn-23 commented on mindmeld/models/taggers/pytorch_crf.py:
```diff
@@ -0,0 +1,407 @@
+        m.bias.data.fill_(0.01)
+
+
+def custom_collate(sequence):
```
Renamed to something more meaningful

vrdn-23 commented on mindmeld/models/taggers/pytorch_crf.py:
```diff
@@ -0,0 +1,407 @@
+        self.num_classes = None
+        self.classes = None
+        self.num_feats = num_feats
+        self.fit_done = False
```
I've renamed it to self.ready. Although I'm not sure if Mindmeld is supposed to pick it up anywhere. This is in the encoder part of the model. Not the actual model itself. I'm not even sure we need it. Hmm

vrdn-23 replied:
So the variable in itself is something that I have leftover from porting my initial code. I don't use to anywhere as a flag or anything to check whether training has been done. I can either completely remove it, or should I be adding a check somewhere? I don't see an equivalent for this in the current crf-suite implementation so I'm not sure if maybe it's needed. Thoughts? @serapio @vembar

vrdn-23 replied:
Resolved offline

tmehlinger commented:
On memory consumption, I'm not terribly concerned. Obviously don't allocate all the way to the moon just because it's convenient and keep it in line with present consumption (or even save memory!). However, we have plenty of flexibility to provide resources if you _need_ memory.

murali1996 commented on mindmeld/models/tagger_models.py:
```diff
@@ -73,12 +73,13 @@ class TaggerModel(Model):
     SEQ_ACCURACY_SCORING = "seq_accuracy"
-    SEQUENCE_MODELS = ["crf"]
+    SEQUENCE_MODELS = ["crf", "torch-crf"]
 
```
Once the old crf codebase is removed, do we intend to remove the "crf" model name or "torch-crf" ? 
Just a thought: I think because there will be a less chance that we create a different backend based crf in near future (i.e. after this PR is merged), it would be a good idea to retain the old name "crf" as this torch based crf is going to be the de facto crf implementation in MindMeld.

vrdn-23 replied:
Raised https://github.com/cisco/mindmeld/issues/416 for this

snow0x2d0 replied:
I'd say just mention the issue number in the comment.

vrdn-23 replied:
Done

snow0x2d0 reviewed (changes requested):
Still looking over the `pytorch_crf.py` file but these are some initial style nitpicks that caught my eye.

murali1996 commented on mindmeld/models/taggers/pytorch_crf.py:
```diff
@@ -0,0 +1,408 @@
+        super().__init__()
+        self.optimizer = None
+        self.encoder = None
+        self.best_model_save_path = os.path.join(USER_CONFIG_DIR, "tmp", str(uuid.uuid4()), "best_crf_model.pt")
```
As dicussed offline, it is better to save model within the `.generated` folder of an app to avoid any unexpected issues while loading back models from disk.

vrdn-23 replied:
Fixed, I think

vrdn-23 replied:
We'll see if something breaks

murali1996 commented on mindmeld/models/taggers/pytorch_crf.py:
```diff
@@ -0,0 +1,408 @@
+
+logger = logging.getLogger(__name__)
+
+DEFAULT_PYTORCH_CRF_ER_CONFIG = {
```
Better to move it to [components/_config.py](https://github.com/cisco/mindmeld/blob/master/mindmeld/components/_config.py) file.

vrdn-23 replied:
Refactored to make it class variables so no longer applies.

murali1996 commented on mindmeld/models/taggers/pytorch_crf.py:
```diff
@@ -0,0 +1,408 @@
+    "epochs": 100,
+    "train_val_split": 0.15,
+    "optimizer_type": "sgd",  # ["sgd", "adam"]
+}
```
Would like to bring up the [comment](https://github.com/cisco/mindmeld/pull/413#discussion_r882145224) about trying to maintain similar terminology across different modules that use pytorch.

vrdn-23 replied:
Renamed to be consistent across NN models

murali1996 commented on mindmeld/models/taggers/pytorch_crf.py:
```diff
@@ -0,0 +1,408 @@
+
+
+def diag_concat_coo_tensors(tensors):
+    assert len(tensors) > 0
```
Might be a good idea to add a `logger.debug(..)` statement here with a brief message.

vrdn-23 replied:
Done

murali1996 commented on mindmeld/models/taggers/pytorch_crf.py:
```diff
@@ -0,0 +1,408 @@
+        self.patience = params.get('patience', DEFAULT_PYTORCH_CRF_ER_CONFIG['patience'])
+        self.epochs = params.get('epochs', DEFAULT_PYTORCH_CRF_ER_CONFIG['epochs'])
+        self.train_val_split = params.get('train_val_split', DEFAULT_PYTORCH_CRF_ER_CONFIG['train_val_split'])
+        self.optimizer_type = params.get('optimizer_type', DEFAULT_PYTORCH_CRF_ER_CONFIG['optimizer_type']).lower()
```
I think it is better to raise an error if user mentions uppercased string like "Adam", "ADAM" , etc. instead of "adam" as we'll only mention the latter in the docs. Implementation-wise, I think this is not an issue here but if there was a scenario where we propagate the params further down into another method and missed to do lowercasing there; debugging will become difficult.

vrdn-23 replied:
Fixed as mentioned

murali1996 commented on mindmeld/models/taggers/pytorch_crf.py:
```diff
@@ -0,0 +1,408 @@
+        return torch.exp(prob).transpose(0, 1)
+
+    def set_params(self, **params):
+        self.feat_type = params.get('feat_type', DEFAULT_PYTORCH_CRF_ER_CONFIG['feat_type']).lower()
```
nit: maybe you can do something like:
```
params = {
    **DEFAULT_PYTORCH_CRF_ER_CONFIG,
    **params
} 
```
and then use the params to initialize all the required vars; for a better readability.

murali1996 commented on mindmeld/models/taggers/pytorch_crf.py:
```diff
@@ -0,0 +1,408 @@
+from random import randint
+
+import numpy as np
+import torch
```
Note for future: Since we plan to include torch as a requirement in MindMeld, some snippet in neural network codebase can be modified to directly `import torch` and use `nn.Module` and other torch functions directly.

vrdn-23 replied:
Raised issue https://github.com/cisco/mindmeld/issues/415 for this!

snow0x2d0 commented on mindmeld/models/taggers/pytorch_crf.py:
```diff
@@ -0,0 +1,408 @@
+            labels = []
+        if fit:
+            if self.feat_extract_type == "dict":
+                comb_dict_list = [x for seq in feat_dicts for x in seq]
```
Nested comprehensions also tend to be difficult to read

vrdn-23 replied:
Changed to itertool.chain()

snow0x2d0 commented on mindmeld/models/taggers/pytorch_crf.py:
```diff
@@ -0,0 +1,408 @@
+        seq_lens = [len(x) for x in feat_dicts]
+        max_seq_len = max(seq_lens)
+
+        for i, (x, y) in enumerate(zip_longest(feat_dicts, labels)):
```
it's unclear from the context what the shape of `x` is

vrdn-23 replied:
Refactored code and added comment to make it more clear

snow0x2d0 commented on mindmeld/models/taggers/pytorch_crf.py:
```diff
@@ -0,0 +1,408 @@
+                                           constant_values=self.pad_index)
+                label_tensor = torch.as_tensor(transformed_label, dtype=torch.long)
+                encoded_labels.append(label_tensor)
+        return (feats, encoded_labels, seq_lens) if encoded_labels else (feats, seq_lens)
```
I'd suggest keeping the shape of the return value the same and just letting the optional value be `None`
```suggestion
        return (feats, seq_lens, encoded_labels)
```

vrdn-23 replied:
Refactored to make it more neat

snow0x2d0 commented on mindmeld/models/taggers/pytorch_crf.py:
```diff
@@ -0,0 +1,408 @@
+        np.random.seed(self.random_state + 2)
+
+    def validate_params(self):
+        if self.optimizer_type not in ["sgd", "adam"]:
+            raise MindMeldError(
+                f"Optimizer type {self.optimizer_type} not supported. Supported options are ['sgd', 'adam']")
+        elif self.feat_type not in ["hash", "dict"]:
+            raise MindMeldError(f"Feature type {self.feat_type} not supported. Supported options are ['hash', 'dict']")
+        elif not 0 < self.train_val_split < 1:
+            raise MindMeldError("Train-dev split should be a value between 0 and 1.")
+        elif not 0 <= self.drop_input < 1:
+            raise MindMeldError("Drop Input should be a value between 0 and 1. (inclusive)")
```
These can just be single `if`'s, since you're raising if they're true.

vrdn-23 replied:
Fixed

snow0x2d0 commented on mindmeld/models/taggers/pytorch_crf.py:
```diff
@@ -0,0 +1,408 @@
+        _patience_counter = 0
+
+        self.build_params(self.encoder.num_feats, self.encoder.num_classes)
+        if self.optimizer_type == "sgd":
+            self.optimizer = optim.SGD(self.parameters(), lr=0.01, momentum=0.9, nesterov=True, weight_decay=1e-5)
+        elif self.optimizer_type == "adam":
+            self.optimizer = optim.Adam(self.parameters(), weight_decay=1e-5)
```
I would just split this into two 'ifs'

vrdn-23 replied:
Done

snow0x2d0 commented on mindmeld/models/taggers/pytorch_crf.py:
```diff
@@ -0,0 +1,408 @@
+                raise MindMeldError(f"{y} should be am integer value.")
+
+    def build_params(self, num_features, num_classes):
+        self.W = nn.Parameter(torch.nn.init.xavier_normal_(torch.empty(size=(num_features, num_classes))),
+                              requires_grad=True)
+        self.b = nn.Parameter(torch.nn.init.constant_(torch.empty(size=(num_classes,)), val=0.01),
+                              requires_grad=True)
+        self.crf_layer = CRF(num_classes, batch_first=True)
+        self.crf_layer.apply(init_weights)
+        self.num_classes = num_classes
```
We should just add these to `__init__` with None as the default value.

snow0x2d0 commented on mindmeld/models/taggers/pytorch_crf.py:
```diff
@@ -0,0 +1,408 @@
+        self.encoder = Encoder(feature_extractor=self.feat_type, num_feats=self.feat_num)
+        stratify_tuples = None
+        if self.stratify:
+            stratify_tuples = [tuple(sorted(list(set(label)))) for label in y]
```
I'd say maybe just pull this out into a simple function where you can provide some comments as to the purpose of each step here.

ritvikshrivastava commented on mindmeld/models/taggers/crf.py:
```diff
@@ -182,6 +183,140 @@ def setup_model(self, config):
+            marginal_tuples.append(query_marginal_tuples)
+        return marginal_tuples
+
+    def predict_proba_distribution(self, examples, config, resources):
```
This implementation in the earlier version of CRF used marginals to get the distribution instead of the actual prob distribution because of limitations with the earlier approach. With the torch-crf module we can change this method to output absolute probabilities instead (@vembar)

vrdn-23 replied:
Created https://github.com/cisco/mindmeld/issues/421 to address this.

ritvikshrivastava commented on mindmeld/models/taggers/pytorch_crf.py:
```diff
@@ -0,0 +1,408 @@
+
+        return marginals_dict
+
+    def predict(self, X):
```
Is this returning the distribution across labels per token or the top label per token? Can you add a docstring specifying that or otherwise?

vrdn-23 replied:
Added docstring

ritvikshrivastava commented on mindmeld/models/taggers/pytorch_crf.py:
```diff
@@ -0,0 +1,408 @@
+                probs = self.compute_marginal_probabilities(inputs, mask).tolist()
+                mask = mask.tolist()
+                # If anyone has any suggestions on a cleaner way to do this, I am all ears!
+                marginals_dict.extend([[dict(zip(self.encoder.classes, token_probs)) \
```
Can we break this nested list comprehension into individual structures? The three level zipping in a single statement might become tedious to follow later

vrdn-23 replied:
Refactored into a more readable format

snow0x2d0 commented on mindmeld/models/taggers/pytorch_crf.py:
```diff
@@ -273,66 +297,68 @@ def compute_marginal_probabilities(self, inputs, mask):
-        self.random_state = params.get('random_state', randint(1, 10000001))
+    def set_params(self, feat_type="hash", feat_num=50000, stratify_train_val_split=True, drop_input=0.2, batch_size=8,
+                   patience=3, number_of_epochs=100, dev_split_ratio=0.2, optimizer="sgd",
+                   random_state=randint(1, 10000001)):
```
Here you'll probably want to set the default of random_state to `None` and then do something like:

```python
random_state = random_state or randint(1, 10000001)
```

As written the randint function will get called when the _class_ is created and then every function call would use the same random value, which I don't think is what you'd want.

vrdn-23 replied:
Fixed

vrdn-23 commented on mindmeld/models/taggers/pytorch_crf.py:
```diff
@@ -0,0 +1,408 @@
+    def __init__(self, feature_extractor="hash", num_feats=50000):
+
+        self.feat_extractor = DictVectorizer(dtype=np.float32) if feature_extractor == "dict" else FeatureHasher(
+            n_features=num_feats, dtype=np.float32)
```
Fixed

vrdn-23 commented on mindmeld/models/taggers/pytorch_crf.py:
```diff
@@ -0,0 +1,408 @@
+                                           constant_values=self.pad_index)
+                label_tensor = torch.as_tensor(transformed_label, dtype=torch.long)
+                encoded_labels.append(label_tensor)
+        return (feats, encoded_labels, seq_lens) if encoded_labels else (feats, seq_lens)
```
Split into different functions resulting in significantly lesser branches

vrdn-23 commented on mindmeld/models/tagger_models.py:
```diff
@@ -246,8 +248,8 @@ def fit(self, examples, labels, params=None):
-            if self._clf.__class__ == LstmModel:
-                raise MindMeldError("The LSTM model does not support cross-validation")
+            if self._clf.__class__ in (LstmModel, TorchCrfTagger):
+                raise MindMeldError(f"The {self._clf.__class__.__name__} model does not support cross-validation")
```
Fixed to suggested change.

vrdn-23 commented on mindmeld/models/taggers/crf.py:
```diff
@@ -182,6 +183,140 @@ def setup_model(self, config):
+        return self
+
+    def set_params(self, **parameters):
+        self._clf = TorchCrfModel()
```
This involves a much larger refactoring across all existing tagger models. I've created https://github.com/cisco/mindmeld/issues/420 to address this.

snow0x2d0 commented on mindmeld/models/taggers/crf.py:
```diff
@@ -182,6 +183,155 @@ def setup_model(self, config):
+            resources (dict): Resources which may be used for this model's feature extraction
+
+        Returns:
+            list of tuples of (mindmeld.core.QueryEntity): a list of predicted labels \
+             with confidence scores
```
A list of tuples doesn't seem to match what's actually being returned here.

vrdn-23 replied:
Updated the docs accordingly

vrdn-23 commented:
Created https://github.com/cisco/mindmeld/issues/424 to address the incremental build failure

vrdn-23 commented on mindmeld/models/helpers.py:
```diff
@@ -534,6 +535,11 @@ def add_resource(func):
+    if isinstance(val, np.generic):
+        return val.item()
+
+
```
Nice catch! Fixed.

tmehlinger commented on mindmeld/models/taggers/crf.py:
```diff
@@ -182,6 +183,155 @@ def setup_model(self, config):
+        for query_index, query_seq in enumerate(seq):
+            tags = []
+            preds = []
+            for i in range(len(query_seq)):
```
You can give a nicer visual cue using `enumerate` and throwing away the value.
```python
for i, _ in enumerate(query_seq):
    ...
```

vrdn-23 replied:
Done

vrdn-23 commented on mindmeld/models/helpers.py:
```diff
@@ -534,6 +535,13 @@ def add_resource(func):
+def np_encoder(val):
+    if isinstance(val, np.generic):
+        return val.item()
+    else:
```
Thanks Travis! Fixed! :)

vrdn-23 commented on mindmeld/models/tagger_models.py:
```diff
@@ -392,13 +396,21 @@ def _dump(self, path):
-                "model_config": self.config
-            })
+            if isinstance(self._clf, TorchCrfTagger):
+                self._clf.dump(path)
```
Done
